src/advanced_unsafety/uninitialized.md

```diff
@@ -0,0 +1,128 @@
+# Uninitialized memory
+
+> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
+>
+> — _Emily Dickinson_
+
+While we have covered [invalid values], there's another thing that behaves a lot like invalid values, but has nothing to do with actual bit patterns: Uninitialized memory.
+
+An easy way to think about uninitialized memory is that there's an additional value (often called `undef` using LLVM's term for it) that does not map to any concrete bit pattern, but can be introduced in abstract in various ways, and makes _most_ values invalid.
+
+If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be "initialized" with no overhead and then written to in parts.
+
+## Sources of uninitialized memory
+
+### `mem::uninitialized()` and `MaybeUninit::assume_init()`
+
+[`mem::uninitialized()`] is a deprecated API that has a very tempting shape, it lets you do things like `let x = mem::uninitialized()` for cases when you want to construct the value in bits. It's basically _always_ UB to use, since it immediately sets `x` to uninitialized memory, which is UB.
+
+Use [`MaybeUninit<T>`] instead.
+
+It is still possible to create uninitialized memory using [`MaybeUninit::assume_init()`] if you have not, in fact, assured that things are initialized.
+
+### Padding
+
+Padding bytes in structs and enums are often but not always uninitialized. This means that treating a struct as a bag of bytes (by, say, treating `&Struct` as `&[u8; size_of::<Struct>()]` and reading from there) is UB even if you don't write invalid values to those bytes, since you are ginning up uninitialized `u8`s.
```

Comment:
I think we can strengthen this statement to "Padding bytes in structs and enums are uninitialized unless proven otherwise". It may also be good to mention how any typed copies (e.g. moves) may set the padding bytes back to `undef`.

The bag of bytes statement is a really good example since I think almost everyone has held that misconception at some point.

"ginning up" => "accessing": if they're reading them then I think we can use a little more straightforward terminology. Maybe "producing" instead to match the terminology from the book better? This section seems like it was originally working with deep reference validity.

Reply:
Yeah so I wanted to bring this up: you've mentioned before that padding bytes can infact be "not uninit", but looking at the UCG glossary there, it seems like they are _always_ uninit.

Reply:
My guess is that we're talking about the same semantics for padding bytes, but this is basically what I mean by "uninitialized unless proven otherwise": https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=262fc5de7ac226e24770f56e1972eafb
In that case, it's always sound to read the padding bytes of the specific `Foo` being passed around by ref. If you have some random `Foo` value then you can do the same thing as long as you don't move it around (albeit that's very hard to avoid).

Reply:
Yeah that's not in concert with what UCG says: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/glossary.md#padding

> Copying Pad ignores the source byte, and writes any value to the target byte. Or, equivalently (in terms of Abstract Machine behavior), copying Pad marks the target byte as uninitialized.

Reply:
We're just creating a reference to some `Foo` though, we're not copying it around. And even if we did copy out of it, the typed copy operation doesn't modify the source bytes so padding would remain unchanged (initialized to zero).

Reply:
we copied bytes out of it though?

Reply:
Yes but we copied `u8`s out, not `Pad`s.

Reply:
Given that there is no way to copy the unnameable Pad type out as itself, I fail to see how the UCG glossary entry makes any sense unless it is talking about cast-and-then-read

Reply:
The way I read it was that `Pad` is only used in the struct definition, and so only performs its special copy behavior when the struct is moved (i.e. as part of the struct). Basically the struct type is valid to _overlay_ on top of the bytes, moving the value out won't modify those original padding bytes, and you can cast back to a `[u8]` and read them. If you copied/moved the value, then the new value _will_ have uninit padding.

Reply:
Ah, fair point. I think we should file an upstream issue with concrete examples.

Reply:
https://github.com/rust-lang/unsafe-code-guidelines/issues/395

Reply:
Added a little "we don't know what the details are" header

Reply:
I got my answer and have reworked the section